_posts/2018-12-05-exploit-model.md

```diff
@@ -0,0 +1,195 @@
+
+```
++--------+              +---------+
+|  p1a   |              |   p2a   |                      APP
++-------------------+   +-------------------+
+|    nabla LibOS    |   |   nabla LibOS     |            LibOS
+|    (code exec)    |   |                   |            
++-------------------+   +-------------------+
+<<<   7 syscalls  >>>   <<<   7 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+For nabla, the attack surface to obtain code exec in the kernel would need to be from these 7 syscalls.
+
+
+
+To summarize, here's a side-by-side comparison of the models:
+
+```
+      gVisor                   NABLA
+      ------                   -----
+
++--------+              
+|  p1a   |                                              APP
++--------+              +---------+
+<<< 300 syscalls  >>>   |   p2a   |          
++-------------------+   +-------------------+
+|      gVisor       |   |   nabla LibOS     |           LibOS
+|   (code exec)     |   |                   |            
++-------------------+   +-------------------+
+<<<  83 syscalls  >>>   <<<   7 syscalls  >>>     <--- Attack surface 
++--------------------------------------------------+
+|                    kernel                        |     KERNEL
++--------------------------------------------------+
+```
+
+
+We note a subtle difference in the way the application links to the LibOS. In gVisor, the LibOS acts as an additional protection boundary which guards[^2] against the inputs that can be invoked with the 83 syscalls.
+
+
+## But it's a memory safe langauge!
+
+Some argue that gVisor is written in golang, and therefore memory safe. Golang is safer, but is not immune to the same class of bugs that we've discussed in this post.
+
+I believe that golang is a safer language from the programming language standpoint (i.e. the common design pattern doesn't allow writing to arbitrary pointers and performing malloc/free operations), but it is hard to guarantee that the implementation is true to the language semantics. Often, bugs in implementation or performance trade-offs (this is rampant in the hardware work with [Spectre](https://meltdownattack.com/) variant bugs) are the result of this.
+
+As a bonus, here are two implementations of memory corruption POCs - one from an implementation bug, and one from a performance design standpoint.
+
+### Implementation bug
+
+The security research and CTF group that I am part of, [Plaid Parliament of Pwning (PPP)](https://pwning.net/), worked on showing a memory corruption attack leading to code exec in the golang playground. You may view the blog post by Alex Reece [here](http://codearcana.com/posts/2013/04/23/exploiting-a-go-binary.html).
```

Comment:
the link to ppp shows invalid certificate...

Reply:
yea... that's how we roll... jk

I think someone is working on redoing the site, and the certs probably are not configured yet... for now I will change it to http.